cmd/snap/cmd_run.go

```diff
@@ -482,29 +473,103 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func getInfoAndApp(snapName, appName string, rev snap.Revision) (*snap.Info, *snap.AppInfo, error) {
+	info, err := getSnapInfo(snapName, rev)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	app := info.Apps[appName]
+	if app == nil {
+		return nil, nil, fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	}
+
+	return info, app, nil
+}
+
+func maybeWaitWhileInhibited(snapName string, appName string) (info *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock, err error) {
+	if features.RefreshAppAwareness.IsEnabled() {
+		flow := newInhibitionFlow(snapName)
+		notified := false
+		notInhibited := func() (err error) {
+			// get updated "current" snap info
+			info, app, err = getInfoAndApp(snapName, appName, snap.R(0))
+			return err
+		}
+		inhibited := func(hint runinhibit.Hint, inhibitInfo *runinhibit.InhibitInfo) (cont bool, err error) {
+			if !notified {
+				info, app, err = getInfoAndApp(snapName, appName, inhibitInfo.Previous)
+				if err != nil {
+					return false, err
+				}
+				// don't wait, continue with old revision
+				if app.IsService() {
+					return true, nil
+				}
+				// don't start flow if we are not inhibited for refresh
+				if hint != runinhibit.HintInhibitedForRefresh {
+					return false, nil
+				}
+				// start notification flow
+				if err := flow.Notify(); err != nil {
+					return true, err
+				}
+				// make sure we call notification flow only once
+				notified = true
+			}
+			return false, nil
+		}
```

Comment:
(later me here) the `inhibited` callback could be called multiple times. Why don't you want to call the notification flow more than once, if it gets re-inhibited again? I assume we want to let the user know somehow that the snap is inhibited again, so they're not sitting there staring at a completed progress bar.

Reply:
You are correct that `inhibited` callback can be called multiple times for the same inhibition, this is why it is only allowed to run once to avoid spamming the user. It is very unlikely that a snap gets refreshed twice in < 1 second, but if it does then it wouldn't hurt if the user kept waiting for the second refresh to complete, I think it might even make sense from a UX perspective to do that. I feel it might be a bit confusing if I get two consecutive flows for two different refreshes from a single snap run.

Thank you for bringing this up, as it will definitely be helpful when doing the RAA UX review. I think these edge cases can be better addressed in the `Inhibition Flows` once a proper UX review is conducted to avoid cluttering snap run logic further

I might be misunderstanding your point, please let me know if I didn't address your questions?

Reply:
Makes sense, thank you! I guess it's a UX question whether it makes more sense for a single flow to run, then wait again while second silently happens, or for two flows to run in this (very unlikely) case. I think I favor the former, at least if the refreshes don't overlap, which I don't think they should. And if refreshes take >= 1 second, which I'm not sure is always the case. If a refresh could take a long time, due to slow network or something, I think there should be more than one flow.